src/cpu/pred/simple_btb.hh

```diff
@@ -52,41 +54,137 @@ namespace gem5
 namespace branch_prediction
 {
 
-class SimpleBTB : public BranchTargetBuffer
+class BTBEntry : public CacheEntry
 {
   public:
-    SimpleBTB(const SimpleBTBParams &params);
+    /** The entry's tag. */
+    Addr tag;
 
-    void memInvalidate() override;
-    bool valid(ThreadID tid, Addr instPC) override;
-    const PCStateBase *lookup(ThreadID tid, Addr instPC,
-                           BranchType type = BranchType::NoBranch) override;
-    void update(ThreadID tid, Addr instPC, const PCStateBase &target_pc,
-                           BranchType type = BranchType::NoBranch,
-                           StaticInstPtr inst = nullptr) override;
-    const StaticInstPtr getInst(ThreadID tid, Addr instPC) override;
+    /** The entry's target. */
+    std::unique_ptr<PCStateBase> target;
 
+    /** The entry's thread id. */
+    ThreadID tid;
 
+    /** Whether or not the entry is valid. */
+    bool valid;
+
+    /** Pointer to the static branch instruction at this address */
+    StaticInstPtr inst;
+
+    /** Update the entry in the BTB */
+    void update(ThreadID _tid,
+                const PCStateBase &_target,
+                StaticInstPtr _inst)
+    {
+        tid = _tid;
+        set(target, _target);
+        inst = _inst;
+        valid = true;
+    }
+
+    /** Match the tag of the BTB entry */
+    bool matchTag(const Addr _tag, ThreadID _tid)
+    {
+        return (isValid() && (tag == _tag) && (tid == _tid));
+    }
+
+    /** Set the tag of the BTB entry */
+    void setTag(Addr _tag) override { tag = _tag; }
+
+    /** Copy constructor */
+    BTBEntry(const BTBEntry &other)
+    {
+        tag   = other.tag;
+        tid   = other.tid;
+        valid = other.valid;
+        inst  = other.inst;
+    }
+
+    /** Default constructor */
+    BTBEntry() : CacheEntry(), tag(0), tid(0), valid(false), inst(nullptr)  {}
+
+    using CacheEntry::matchTag;
```

Comment:
Why do we allow both to be used? Don´t we always need to know the thread id? Also, is the thread id part of the tag (i.e., can we have multiple threads using the same tag)? If so, then the API of the cache should use a Tag type, not only a regular addr.